File: src/examples/go/long-agents/self_improving_lab.go
// ax-example:start
// title: Go Self-Improving Lab Agent
// group: long-agents
// description: A many-tool agent that runs experiments, grades them against a rubric with an independent verifier, and distills verified rules into memory -- iterating until the rubric passes.
// provider: openai
// env: OPENAI_API_KEY, OPENAI_APIKEY
// level: advanced
// order: 40
// ax-example:end
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"os"
	"regexp"
	"strings"
	"time"

	ax "github.com/ax-llm/ax/packages/go"
	axgoja "github.com/ax-llm/ax/packages/go/runtime/goja"
)

func openAIClient() *ax.OpenAICompatibleClient {
	apiKey := os.Getenv("OPENAI_API_KEY")
	if apiKey == "" {
		apiKey = os.Getenv("OPENAI_APIKEY")
	}
	if apiKey == "" {
		panic("Set OPENAI_API_KEY or OPENAI_APIKEY to run this example.")
	}
	model := os.Getenv("AX_OPENAI_MODEL")
	if model == "" {
		model = "gpt-4o-mini"
	}
	return ax.NewOpenAICompatibleClient(map[string]ax.Value{"api_key": apiKey, "model": model, "model_config": ax.Object("temperature", 0)})
}

func printJSON(value ax.Value) {
	data, err := json.MarshalIndent(value, "", "  ")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(data))
}

func asMap(value ax.Value) map[string]ax.Value {
	if out, ok := value.(map[string]ax.Value); ok {
		return out
	}
	return map[string]ax.Value{}
}

func asString(value ax.Value) string {
	if value == nil {
		return ""
	}
	if s, ok := value.(string); ok {
		return s
	}
	return fmt.Sprint(value)
}

// ---------------------------------------------------------------------------
// The "lab": a deterministic black-box experiment. It scores an ETL config plan
// against a hidden ideal and returns, for any failing check, the exact fix --
// so the agent can converge by following the feedback, not by being told.
// ---------------------------------------------------------------------------
var checks = []string{"no-nulls", "no-duplicates", "numeric-types", "trimmed-strings", "outliers-handled"}

var remedies = map[string]string{
	"no-nulls":         "set nullPolicy=impute (or nullPolicy=drop)",
	"no-duplicates":    "set dedup=on",
	"numeric-types":    "set coerceTypes=on",
	"trimmed-strings":  "set trim=on",
	"outliers-handled": "set outlier=clip (or outlier=winsorize)",
}

var flagRe = regexp.MustCompile(`([a-z]+)\s*=\s*([a-z0-9]+)`)

func runInSandbox(plan string) ax.Value {
	flags := map[string]string{}
	for _, m := range flagRe.FindAllStringSubmatch(strings.ToLower(plan), -1) {
		flags[m[1]] = m[2]
	}
	ok := map[string]bool{
		"no-nulls":         flags["nullpolicy"] == "impute" || flags["nullpolicy"] == "drop",
		"no-duplicates":    flags["dedup"] == "on",
		"numeric-types":    flags["coercetypes"] == "on",
		"trimmed-strings":  flags["trim"] == "on",
		"outliers-handled": flags["outlier"] == "clip" || flags["outlier"] == "winsorize",
	}
	passed := []ax.Value{}
	failed := []ax.Value{}
	for _, c := range checks {
		if ok[c] {
			passed = append(passed, c)
		} else {
			failed = append(failed, ax.Object("check", c, "fix", remedies[c]))
		}
	}
	score := math.Round(float64(len(passed))/float64(len(checks))*100) / 100
	return ax.Object(
		"score", score,
		"solved", len(passed) == len(checks),
		"passed", ax.Array(passed...),
		"failed", ax.Array(failed...),
		"logs", fmt.Sprintf("%d/%d checks passed", len(passed), len(checks)),
	)
}

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Minute)
	defer cancel()
	client := openAIClient()

	// An independent verifier -- a separate ax() program, not the agent grading itself.
	verifier := ax.NewAx(
		"rubric:string, evidence:json -> passed:boolean, feedback:string, missing:string[]",
		map[string]ax.Value{"instruction": "You are an independent rubric grader, not a self-critique. Pass only when the evidence clearly satisfies every part of the rubric."},
	)

	// In-memory rule store. Verified, reusable rules go here -- not raw failure notes.
	memoryStore := map[string]string{}
	memoryOrder := []string{}
	storeRule := func(key, value string) {
		if _, seen := memoryStore[key]; !seen {
			memoryOrder = append(memoryOrder, key)
		}
		memoryStore[key] = value
	}

	runtime := axgoja.NewRuntime(
		axgoja.WithCallable("runExperiment", func(p ax.Value) (ax.Value, error) {
			return runInSandbox(asString(asMap(p)["plan"])), nil
		}),
		axgoja.WithCallable("listChecks", func(p ax.Value) (ax.Value, error) {
			out := []ax.Value{}
			for _, c := range checks {
				out = append(out, c)
			}
			return ax.Array(out...), nil
		}),
		axgoja.WithCallable("grade", func(p ax.Value) (ax.Value, error) {
			m := asMap(p)
			evidence := m["evidence"]
			if evidence == nil {
				evidence = ax.Array()
			}
			return verifier.Forward(ctx, client, map[string]ax.Value{"rubric": asString(m["rubric"]), "evidence": evidence}, nil)
		}),
		axgoja.WithCallable("recallRules", func(p ax.Value) (ax.Value, error) {
			t := strings.ToLower(asString(asMap(p)["topic"]))
			words := strings.Fields(t)
			out := []ax.Value{}
			for _, k := range memoryOrder {
				match := strings.Contains(k, t)
				if !match {
					for _, w := range words {
						if strings.Contains(k, w) {
							match = true
							break
						}
					}
				}
				if match {
					out = append(out, memoryStore[k])
				}
			}
			return ax.Array(out...), nil
		}),
		axgoja.WithCallable("remember", func(p ax.Value) (ax.Value, error) {
			m := asMap(p)
			rule := asString(m["rule"])
			key := strings.ToLower(rule)
			if len(key) > 48 {
				key = key[:48]
			}
			storeRule(key, fmt.Sprintf("%s :: %s", rule, asString(m["evidence"])))
			return ax.Object("stored", true, "total", len(memoryStore)), nil
		}),
	)

	spec := func(name, description string, props map[string]ax.Value, required ...string) ax.Value {
		parameters := ax.Object("type", "object", "properties", props)
		if len(required) > 0 {
			reqd := []ax.Value{}
			for _, r := range required {
				reqd = append(reqd, r)
			}
			parameters["required"] = ax.Array(reqd...)
		}
		return ax.Object("name", name, "description", description, "parameters", parameters)
	}

	selfImproving := ax.NewAgent(
		`goal:string, rubric:string -> answer:string, experiments:string[] "Plans tried, in order", learnedRules:string[]`,
		map[string]ax.Value{
			"contextFields": ax.Array(),
			"functions": ax.Array(
				spec("runExperiment", "Apply an ETL config plan; returns score, solved, passed[], failed[{check,fix}], logs. Pass an empty plan to discover the fixes.", map[string]ax.Value{"plan": ax.Object("type", "string")}, "plan"),
				spec("listChecks", "List the data-quality checks the experiment evaluates.", map[string]ax.Value{}),
				spec("grade", "Independent rubric grader. Pass only when the evidence meets the rubric.", map[string]ax.Value{"rubric": ax.Object("type", "string"), "evidence": ax.Object("type", "array", "items", ax.Object("type", "string"))}, "rubric", "evidence"),
				spec("recallRules", "Recall verified rules relevant to a topic.", map[string]ax.Value{"topic": ax.Object("type", "string")}, "topic"),
				spec("remember", "Store a verified, reusable rule (the rule, not raw notes).", map[string]ax.Value{"rule": ax.Object("type", "string"), "evidence": ax.Object("type", "string")}, "rule", "evidence"),
			),
			"contextPolicy": ax.Object("preset", "adaptive", "budget", "balanced"),
			"executorOptions": ax.Object("description", strings.Join([]string{
				"Use the tools -- do not answer from your own knowledge.",
				"1. recallRules('etl data quality') to reuse anything already learned.",
				"2. runExperiment('') once to see every failing check and its fix.",
				"3. Build a plan applying all the fixes, then runExperiment again. Repeat until solved is true.",
				"4. grade the passing evidence against the rubric.",
				"5. For each check you fixed, remember(rule, evidence).",
				"6. Then return the answer, the plans you tried, and the learned rules.",
			}, "\n")),
			"runtime": ax.Object("language", "JavaScript"),
		},
	)

	result, err := selfImproving.Forward(
		ctx,
		client,
		map[string]ax.Value{
			"goal":   "Find an ETL config plan that cleans the dirty dataset so every data-quality check passes.",
			"rubric": "All five checks (no-nulls, no-duplicates, numeric-types, trimmed-strings, outliers-handled) must pass, i.e. score 1.0.",
		},
		map[string]ax.Value{"runtime": runtime, "max_actor_steps": 18},
	)
	if err != nil {
		panic(err)
	}

	printJSON(result)

	// Persist the agent's verified rules so a future run's recall reuses them.
	if learned, ok := asMap(result)["learnedRules"].([]ax.Value); ok {
		for _, rule := range learned {
			r := asString(rule)
			key := strings.ToLower(r)
			if len(key) > 48 {
				key = key[:48]
			}
			storeRule(key, r)
		}
	}
	fmt.Printf("\nMemory now holds %d rule(s) for next time.\n", len(memoryStore))
}

File: src/examples/cpp/long-agents/self_improving_lab.cpp
// ax-example:start
// title: C++ Self-Improving Lab Agent
// group: long-agents
// description: A many-tool agent that runs experiments, grades them against a rubric with an independent verifier, and distills verified rules into memory -- iterating until the rubric passes.
// provider: openai
// env: OPENAI_API_KEY, OPENAI_APIKEY
// level: advanced
// order: 40
// ax-example:end
#include "axllm/axllm.hpp"
#include "axllm/runtime/quickjs/quickjs_runtime.hpp"
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <iostream>
#include <map>
#include <regex>
#include <sstream>
#include <string>
#include <vector>

// ---------------------------------------------------------------------------
// The "lab": a deterministic black-box experiment. It scores an ETL config plan
// against a hidden ideal and returns, for any failing check, the exact fix --
// so the agent can converge by following the feedback, not by being told.
// ---------------------------------------------------------------------------
static const std::vector<std::string> CHECKS = {
    "no-nulls", "no-duplicates", "numeric-types", "trimmed-strings", "outliers-handled"};

static std::string remedy_for(const std::string& check) {
  if (check == "no-nulls") return "set nullPolicy=impute (or nullPolicy=drop)";
  if (check == "no-duplicates") return "set dedup=on";
  if (check == "numeric-types") return "set coerceTypes=on";
  if (check == "trimmed-strings") return "set trim=on";
  if (check == "outliers-handled") return "set outlier=clip (or outlier=winsorize)";
  return "";
}

static std::string to_lower(std::string s) {
  std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::tolower(c); });
  return s;
}

static axllm::Value run_in_sandbox(const std::string& plan) {
  std::string lowered = to_lower(plan);
  std::map<std::string, std::string> flags;
  std::regex re("([a-z]+)\\s*=\\s*([a-z0-9]+)");
  for (std::sregex_iterator it(lowered.begin(), lowered.end(), re), end; it != end; ++it) {
    flags[(*it)[1].str()] = (*it)[2].str();
  }
  auto flag = [&](const std::string& key) -> std::string {
    auto found = flags.find(key);
    return found == flags.end() ? "" : found->second;
  };

  std::map<std::string, bool> ok = {
      {"no-nulls", flag("nullpolicy") == "impute" || flag("nullpolicy") == "drop"},
      {"no-duplicates", flag("dedup") == "on"},
      {"numeric-types", flag("coercetypes") == "on"},
      {"trimmed-strings", flag("trim") == "on"},
      {"outliers-handled", flag("outlier") == "clip" || flag("outlier") == "winsorize"},
  };

  axllm::Value passed = axllm::Value::array();
  axllm::Value failed = axllm::Value::array();
  int passed_count = 0;
  for (const auto& c : CHECKS) {
    if (ok[c]) {
      axllm::Core::append(passed, c);
      ++passed_count;
    } else {
      axllm::Core::append(failed, axllm::object({{"check", c}, {"fix", remedy_for(c)}}));
    }
  }
  double score = std::round((static_cast<double>(passed_count) / CHECKS.size()) * 100.0) / 100.0;
  std::ostringstream logs;
  logs << passed_count << "/" << CHECKS.size() << " checks passed";
  return axllm::object({
      {"score", score},
      {"solved", passed_count == static_cast<int>(CHECKS.size())},
      {"passed", passed},
      {"failed", failed},
      {"logs", logs.str()},
  });
}

static axllm::Value fn_spec(const std::string& name, const std::string& description,
                           axllm::Value props, axllm::Value required = axllm::Value()) {
  axllm::Value parameters = axllm::object({{"type", "object"}, {"properties", props}});
  if (!required.is_null()) axllm::Core::set(parameters, "required", required);
  return axllm::object({{"name", name}, {"description", description}, {"parameters", parameters}});
}

int main() {
  const char* key = std::getenv("OPENAI_API_KEY");
  if (key == nullptr || std::string(key).empty()) key = std::getenv("OPENAI_APIKEY");
  if (key == nullptr || std::string(key).empty()) {
    std::cerr << "Set OPENAI_API_KEY or OPENAI_APIKEY to run this example.\n";
    return 2;
  }
  const char* model = std::getenv("AX_OPENAI_MODEL");
  axllm::OpenAICompatibleClient client(axllm::object({
      {"api_key", key},
      {"model", model == nullptr || std::string(model).empty() ? "gpt-5.4-mini" : model},
      {"model_config", axllm::object({{"temperature", 0}})},
  }));

  // An independent verifier -- a separate ax() program, not the agent grading itself.
  auto verifier = axllm::ax("rubric:string, evidence:json -> passed:boolean, feedback:string, missing:string[]");
  verifier.set_instruction(
      "You are an independent rubric grader, not a self-critique. Pass only when the evidence clearly satisfies every part of the rubric.");

  // In-memory rule store. Verified, reusable rules go here -- not raw failure notes.
  std::map<std::string, std::string> memory_store;

  axllm::runtime::quickjs::QuickJsCodeRuntime runtime;
  runtime
      .register_callable("runExperiment", [](axllm::Value p) -> axllm::Value {
        return run_in_sandbox(axllm::display(axllm::Core::get(p, "plan", "")));
      })
      .register_callable("listChecks", [](axllm::Value) -> axllm::Value {
        axllm::Value out = axllm::Value::array();
        for (const auto& c : CHECKS) axllm::Core::append(out, c);
        return out;
      })
      .register_callable("grade", [&client, &verifier](axllm::Value p) -> axllm::Value {
        return verifier.forward(
            client,
            axllm::object({
                {"rubric", axllm::Core::get(p, "rubric", "")},
                {"evidence", axllm::Core::get(p, "evidence", axllm::Value::array())},
            }));
      })
      .register_callable("recall", [&memory_store](axllm::Value p) -> axllm::Value {
        std::string topic = to_lower(axllm::display(axllm::Core::get(p, "topic", "")));
        std::vector<std::string> words;
        std::istringstream ss(topic);
        std::string word;
        while (ss >> word) words.push_back(word);
        axllm::Value out = axllm::Value::array();
        for (const auto& entry : memory_store) {
          bool match = !topic.empty() && entry.first.find(topic) != std::string::npos;
          for (const auto& w : words) {
            if (!w.empty() && entry.first.find(w) != std::string::npos) match = true;
          }
          if (match) axllm::Core::append(out, entry.second);
        }
        return out;
      })
      .register_callable("remember", [&memory_store](axllm::Value p) -> axllm::Value {
        std::string rule = axllm::display(axllm::Core::get(p, "rule", ""));
        std::string evidence = axllm::display(axllm::Core::get(p, "evidence", ""));
        std::string key = to_lower(rule).substr(0, 48);
        memory_store[key] = rule + " :: " + evidence;
        return axllm::object({{"stored", true}, {"total", static_cast<double>(memory_store.size())}});
      });

  auto self_improving = axllm::agent(
      "goal:string, rubric:string -> answer:string, experiments:string[] \"Plans tried, in order\", learnedRules:string[]",
      axllm::object({
          {"contextFields", axllm::array({})},
          {"functions", axllm::array({
              fn_spec("runExperiment",
                      "Apply an ETL config plan; returns score, solved, passed[], failed[{check,fix}], logs. Pass an empty plan to discover the fixes.",
                      axllm::object({{"plan", axllm::object({{"type", "string"}})}}),
                      axllm::array({"plan"})),
              fn_spec("listChecks", "List the data-quality checks the experiment evaluates.", axllm::object({})),
              fn_spec("grade",
                      "Independent rubric grader. Pass only when the evidence meets the rubric.",
                      axllm::object({
                          {"rubric", axllm::object({{"type", "string"}})},
                          {"evidence", axllm::object({{"type", "array"}, {"items", axllm::object({{"type", "string"}})}})},
                      }),
                      axllm::array({"rubric", "evidence"})),
              fn_spec("recall", "Recall verified rules relevant to a topic.",
                      axllm::object({{"topic", axllm::object({{"type", "string"}})}}),
                      axllm::array({"topic"})),
              fn_spec("remember", "Store a verified, reusable rule (the rule, not raw notes).",
                      axllm::object({
                          {"rule", axllm::object({{"type", "string"}})},
                          {"evidence", axllm::object({{"type", "string"}})},
                      }),
                      axllm::array({"rule", "evidence"})),
          })},
          {"contextPolicy", axllm::object({{"preset", "adaptive"}, {"budget", "balanced"}})},
          {"executorOptions", axllm::object({
              {"description",
               std::string("Use the tools -- do not answer from your own knowledge.\n") +
                   "1. recall('etl data quality') to reuse anything already learned.\n" +
                   "2. runExperiment('') once to see every failing check and its fix.\n" +
                   "3. Build a plan applying all the fixes, then runExperiment again. Repeat until solved is true.\n" +
                   "4. grade the passing evidence against the rubric.\n" +
                   "5. For each check you fixed, remember(rule, evidence).\n" +
                   "6. Then return the answer, the plans you tried, and the learned rules."},
          })},
          {"runtime", axllm::object({{"language", "JavaScript"}})},
      }));

  axllm::Value result = self_improving.forward(
      client,
      axllm::object({
          {"goal", "Find an ETL config plan that cleans the dirty dataset so every data-quality check passes."},
          {"rubric", "All five checks (no-nulls, no-duplicates, numeric-types, trimmed-strings, outliers-handled) must pass, i.e. score 1.0."},
      }),
      axllm::object({{"runtime", axllm::Core::code_runtime_ref(runtime)}, {"max_actor_steps", 18}}));

  std::cout << axllm::stringify(result) << "\n";

  // Persist the agent's verified rules so a future run's recall reuses them.
  axllm::Value learned = axllm::Core::get(result, "learnedRules", axllm::Value::array());
  for (const auto& rule : axllm::Core::iter(learned)) {
    std::string text = axllm::display(rule);
    memory_store[to_lower(text).substr(0, 48)] = text;
  }
  std::cout << "\nMemory now holds " << memory_store.size() << " rule(s) for next time.\n";
}
